examples/chip-tool/commands/common/CHIPCommand.cpp

```diff
@@ -29,17 +29,91 @@
 
 using DeviceControllerFactory = chip::Controller::DeviceControllerFactory;
 
+constexpr const char kCommissionerAlpha[] = "alpha";
+constexpr const char kCommissionerBeta[]  = "beta";
+constexpr const char kCommissionerGamma[] = "gamma";
+
+constexpr chip::FabricId kCommissionerAlphaFabricId = 1;
+constexpr chip::FabricId kCommissionerBetaFabricId  = 2;
+constexpr chip::FabricId kCommissionerGammaFabricId = 3;
+
 CHIP_ERROR CHIPCommand::Run()
 {
 #if CHIP_DEVICE_LAYER_TARGET_LINUX && CHIP_DEVICE_CONFIG_ENABLE_CHIPOBLE
     // By default, Linux device is configured as a BLE peripheral while the controller needs a BLE central.
     ReturnLogErrorOnFailure(chip::DeviceLayer::Internal::BLEMgrImpl().ConfigureBle(0, true));
 #endif
 
-    ReturnLogErrorOnFailure(mStorage.Init());
-    ReturnLogErrorOnFailure(mOpCredsIssuer.Initialize(mStorage));
-    ReturnLogErrorOnFailure(mFabricStorage.Initialize(&mStorage));
+    ReturnLogErrorOnFailure(mDefaultStorage.Init());
+    ReturnLogErrorOnFailure(mFabricStorage.Initialize(&mDefaultStorage));
+
+    chip::Controller::FactoryInitParams factoryInitParams;
+    factoryInitParams.fabricStorage = &mFabricStorage;
+    factoryInitParams.listenPort    = static_cast<uint16_t>(mDefaultStorage.GetListenPort() + CurrentCommissionerIndex());
+    ReturnLogErrorOnFailure(DeviceControllerFactory::GetInstance().Init(factoryInitParams));
+
+    ReturnLogErrorOnFailure(InitializeCommissioner(CurrentCommissionerName(), CurrentCommissionerIndex()));
+
+    chip::DeviceLayer::PlatformMgr().ScheduleWork(RunQueuedCommand, reinterpret_cast<intptr_t>(this));
+    ReturnLogErrorOnFailure(StartWaiting(GetWaitDuration()));
+
+    Shutdown();
+
+    //
+    // We can call DeviceController::Shutdown() safely without grabbing the stack lock
+    // since the CHIP thread and event queue have been stopped, preventing any thread
+    // races.
+    //
+    ReturnLogErrorOnFailure(ShutdownCommissioner(CurrentCommissionerName()));
+
+    return CHIP_NO_ERROR;
+}
 
+std::string CHIPCommand::CurrentCommissionerName()
+{
+    std::string name = mCommissionerName.HasValue() ? mCommissionerName.Value() : kCommissionerAlpha;
+    if (name.compare(kCommissionerAlpha) != 0 && name.compare(kCommissionerBeta) != 0 && name.compare(kCommissionerGamma) != 0)
+    {
+        ChipLogError(chipTool, "Unknown commissioner name: %s. Supported names are [%s, %s, %s]", name.c_str(), kCommissionerAlpha,
+                     kCommissionerBeta, kCommissionerGamma);
+        chipDie();
+    }
+
+    return name;
+}
+
+uint16_t CHIPCommand::CurrentCommissionerIndex()
+{
+    std::string name = CurrentCommissionerName();
+    if (name.compare(kCommissionerAlpha) == 0)
+    {
+        return kCommissionerAlphaFabricId;
+    }
+    else if (name.compare(kCommissionerBeta) == 0)
+    {
+        return kCommissionerBetaFabricId;
+    }
+    else if (name.compare(kCommissionerGamma) == 0)
+    {
+        return kCommissionerGammaFabricId;
+    }
+
+    return 0;
```

Comment:
it is unclear here if 0 is a valid index or not.

Please use VerifyOrDie or a constant or at least log that index is not valid.